File: src/main/java/net/felsstudio/fels/parser/ast/ResultVisitor.java
package main.java.net.felsstudio.fels.parser.ast;

import java.io.IOException;

/**
 *
 * @author FelekDevYT,FelsStudio
 * @param <R> the result type
 * @param <T> the type if the input
 */
public interface ResultVisitor<R, T> {
    
    R visit(ArrayExpression s, T t) throws IOException, InterruptedException;
    R visit(AssignmentExpression s, T t) throws IOException, InterruptedException;
    R visit(BinaryExpression s, T t) throws IOException, InterruptedException;
    R visit(BlockStatement s, T t) throws IOException, InterruptedException;
    R visit(BreakStatement s, T t);
    R visit(ClassDeclarationStatement s, T t) throws IOException, InterruptedException;
    R visit(ConditionalExpression s, T t) throws IOException, InterruptedException;
    R visit(ContainerAccessExpression s, T t) throws IOException, InterruptedException;
    R visit(ContinueStatement s, T t);
    R visit(DoWhileStatement s, T t) throws IOException, InterruptedException;
    R visit(DestructuringAssignmentStatement s, T t) throws IOException, InterruptedException;
    R visit(ForStatement s, T t) throws IOException, InterruptedException;
    R visit(ForeachArrayStatement s, T t) throws IOException, InterruptedException;
    R visit(ForeachMapStatement s, T t) throws IOException, InterruptedException;
    R visit(FunctionDefineStatement s, T t) throws IOException, InterruptedException;
    R visit(FunctionReferenceExpression s, T t);
    R visit(ExprStatement s, T t) throws IOException, InterruptedException;
    R visit(FunctionalExpression s, T t) throws IOException, InterruptedException;
    R visit(IfStatement s, T t) throws IOException, InterruptedException;
    R visit(ImportStatement s, T t) throws IOException, InterruptedException;
    R visit(MapExpression s, T t) throws IOException, InterruptedException;
    R visit(MatchExpression s, T t) throws IOException, InterruptedException;
    R visit(ObjectCreationExpression s, T t) throws IOException, InterruptedException;
    R visit(PrintStatement s, T t) throws IOException, InterruptedException;
    R visit(PrintlnStatement s, T t) throws IOException, InterruptedException;
    R visit(ReturnStatement s, T t) throws IOException, InterruptedException;
    R visit(TernaryExpression s, T t) throws IOException, InterruptedException;
    R visit(UnaryExpression s, T t) throws IOException, InterruptedException;
    R visit(ValueExpression s, T t) throws IOException, InterruptedException;
    R visit(VariableExpression s, T t);
    R visit(WhileStatement s, T t) throws IOException, InterruptedException;
    R visit(UsingStatement s, T t) throws IOException, InterruptedException;
}

File: src/main/java/net/felsstudio/fels/parser/ast/ResultVisitor.kt
package main.java.net.felsstudio.fels.parser.ast

/**
 *
 * @author felek
 */
interface ResultVisitor<R, T> {
    fun visit(s: ArrayExpression?, t: T): R
    fun visit(s: AssignmentExpression?, t: T): R
    fun visit(s: BinaryExpression?, t: T): R
    fun visit(s: BlockStatement?, t: T): R
    fun visit(s: BreakStatement?, t: T): R
    fun visit(s: ConditionalExpression?, t: T): R
    fun visit(s: ContainerAccessExpression?, t: T): R
    fun visit(s: ContinueStatement?, t: T): R
    fun visit(s: DoWhileStatement?, t: T): R
    fun visit(s: DestructuringAssignmentStatement?, t: T): R
    fun visit(s: ForStatement?, t: T): R
    fun visit(s: ForeachArrayStatement?, t: T): R
    fun visit(s: ForeachMapStatement?, t: T): R
    fun visit(s: FunctionDefineStatement?, t: T): R
    fun visit(s: FunctionReferenceExpression?, t: T): R
    fun visit(s: ExprStatement?, t: T): R
    fun visit(s: FunctionalExpression?, t: T): R
    fun visit(s: IfStatement?, t: T): R
    fun visit(s: ImportStatement?, t: T): R
    fun visit(s: MapExpression?, t: T): R
    fun visit(s: MatchExpression?, t: T): R
    fun visit(s: PrintStatement?, t: T): R
    fun visit(s: PrintlnStatement?, t: T): R
    fun visit(s: ReturnStatement?, t: T): R
    fun visit(s: TernaryExpression?, t: T): R
    fun visit(s: UnaryExpression?, t: T): R
    fun visit(s: ValueExpression?, t: T): R
    fun visit(s: VariableExpression?, t: T): R
    fun visit(s: WhileStatement?, t: T): R
    fun visit(s: UsingStatement?, t: T): R
    fun visit(s: ClassDeclarationStatement?, t: T): R
    fun visit(objectCreationExpression: ObjectCreationExpression?, t: T): R
    fun <T1, R1> visit(s: RebreakStatement, t: T1): R1
}
